File: backend/app/schemas/analytics_mgmt.py
from typing import Optional, Any, Dict
from datetime import datetime
from pydantic import BaseModel

# --- Report Template Schemas ---

class ReportTemplateBase(BaseModel):
    name: str
    description: Optional[str] = None
    config_json: Dict[str, Any]
    is_public: bool = False

class ReportTemplateCreate(ReportTemplateBase):
    pass

class ReportTemplateUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    config_json: Optional[Dict[str, Any]] = None
    is_public: Optional[bool] = None

class ReportTemplateRead(ReportTemplateBase):
    id: str
    account_id: str
    created_by_user_id: Optional[str]
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True

# --- Custom Metrics Schemas ---

class CustomMetricBase(BaseModel):
    name: str
    description: Optional[str] = None
    formula: str
    format: str = "number"  # number, currency, percent, time

class CustomMetricCreate(CustomMetricBase):
    pass

class CustomMetricUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    formula: Optional[str] = None
    format: Optional[str] = None

class CustomMetricRead(CustomMetricBase):
    id: str
    account_id: str
    created_by_user_id: Optional[str]
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True

File: frontend/src/lib/analytics-mgmt.ts
import { apiFetch } from "./api-client";

// --- Interfaces ---

export interface ReportTemplate {
    id: string;
    account_id: string;
    name: string;
    description?: string;
    config_json: Record<string, any>;
    is_public: boolean;
    created_by_user_id?: string;
    created_at: string;
    updated_at?: string;
}

export interface ReportTemplateCreate {
    name: string;
    description?: string;
    config_json: Record<string, any>;
    is_public: boolean;
}

export interface CustomMetric {
    id: string;
    account_id: string;
    name: string;
    description?: string;
    formula: string;
    format: string; // currency, percentage, number, duration
    created_by_user_id?: string;
    created_at: string;
    updated_at?: string;
}

export interface CustomMetricCreate {
    name: string;
    description?: string;
    formula: string;
    format: string;
}

// --- Service Methods ---

// Report Templates
export async function getTemplates(): Promise<ReportTemplate[]> {
    const res = await apiFetch<ReportTemplate[]>("/analytics/templates");
    return res || [];
}

export async function createTemplate(data: ReportTemplateCreate): Promise<ReportTemplate> {
    const res = await apiFetch<ReportTemplate>("/analytics/templates", {
        method: "POST",
        body: JSON.stringify(data),
    });
    if (!res) throw new Error("Failed to create template");
    return res;
}

export async function deleteTemplate(id: string): Promise<void> {
    await apiFetch<void>(`/analytics/templates/${id}`, {
        method: "DELETE",
    });
}

// Custom Metrics
export async function getCustomMetrics(): Promise<CustomMetric[]> {
    const res = await apiFetch<CustomMetric[]>("/analytics/custom-metrics");
    return res || [];
}

export async function createCustomMetric(data: CustomMetricCreate): Promise<CustomMetric> {
    const res = await apiFetch<CustomMetric>("/analytics/custom-metrics", {
        method: "POST",
        body: JSON.stringify(data),
    });
    if (!res) throw new Error("Failed to create custom metric");
    return res;
}

export async function deleteCustomMetric(id: string): Promise<void> {
    await apiFetch<void>(`/analytics/custom-metrics/${id}`, {
        method: "DELETE",
    });
}
